pkg/cmd/roachtest/roachtestutil/mixedversion/helper.go

```diff
@@ -26,73 +26,191 @@ import (
 	"github.com/cockroachdb/errors"
 )
 
-// Helper is the struct passed to `stepFunc`s (user-provided or
-// implemented by the framework) that provides helper functions that
-// mixed-version tests can use.
-type Helper struct {
-	Context *Context
+type (
+	// ServiceDescriptor encapsulates the information about where a
+	// service (system tenant or otherwise) is running.
+	ServiceDescriptor struct {
+		// Name is the name of the service ("system" for the system
+		// tenant, or the tenant name otherwise.)
+		Name string
 
-	ctx context.Context
-	// bgCount keeps track of the number of background tasks started
-	// with `helper.Background()`. The counter is used to generate
-	// unique log file names.
-	bgCount    int64
-	runner     *testRunner
-	stepLogger *logger.Logger
+		// Nodes is the set of nodes in the cluster where clients can
+		// connect to that service.
+		Nodes option.NodeListOption
```

Comment:
Remind me if we were planning to support multiple tenants on the same node? It doesn't appear to be the case here, so we should maybe clarify that in comments.

Reply:
> support multiple tenants on the same node?

You mean external process, right? That's currently not supported due to the lack of dynamic admin UI ports (but Bhaskar's recent work will change that). So roachprod itself won't have that limitation.

As far as roachtest is concerned, I'm currently leaning on the direction of _not_ co-locating tenants (or even KV + tenants) on the same node. There would be a lack of performance isolation and more generally, seems like a contrived deployment model that is not very realistic.

We could still consider it as a tradeoff if we wanted to explicitly test a large number of tenants, where 1 VM per tenant is prohibitively expensive.

> It doesn't appear to be the case here

Why do you say so? I don't think the code currently requires that -- it's more about how the planner decides to set up these tenants.

Reply:
> Why do you say so? I don't think the code currently requires that -- it's more about how the planner decides to set up these tenants.

If we have multiple tenants co-located on the same node, then the current addressing scheme, i.e., `Nodes` doesn't suffice, right?

Reply:
> then the current addressing scheme, i.e., Nodes doesn't suffice, right?

We use a `ServiceDescriptor` to address services (instead of just nodes). Nodes + service name is enough since we that's all we need using roachprod's service discovery.